Scoped storage implementation of Android 11+ and also bump older vers…

Xavron opened the pull request:
…ions onto it if they have issues with the old code.

**All final tests:**

  [x] == passed
  [] == not tested; not looked into
  [*] == couldn't connect on this device for unknown reason and not debugging this; not tested

    Continual Backups 1GB-81GB:
	Android 12 internal [x]
	Android 12 sd card  [x]
	-
	Android 8 internal  [x]
	Android 8 sd card   [x]
	-
	Android 7 internal  [*]
	Android 7 sd card   [*]
	-
	Android 6 internal  [x]
	Android 6 sd card   []

    Full restore of multiple backups 1GB-81GB:
	Android 12 internal [x]
	Android 12 sd card  [x]
	-
	Android 8 internal  [x]
	Android 8 sd card   [x]
	-
	Android 7 internal  [*]
	Android 7 sd card   [*]
	-
	Android 6 internal  []
	Android 6 sd card   []

    Filezilla:
	Android 12 internal [x] 
	Android 12 sd card  [x] 
	-
	Android 8 internal  [x]
	Android 8 sd card   [x]
	-
	Android 7 internal  [*]
	Android 7 sd card   [*]
	-
	Android 6 internal  [x]
	Android 6 sd card   []

    Other:
	Test backup and FileZilla after app reinstall [x] <- write external storage is a problem here (see known issues).
	Test older File code execution is still working + clean install [x] <- tested on Android 6 device.

    Successful tests include:
	movement
	creations
	deletions
	transfer up
	transfer down
	change of target location without restart of app
	target in sub folders
	target at picker location
	multi-clients (running at the same time and at separate times)
	multi-clients supplied different sub folders
	user added chroot
	picker
	sd card
	internal
	various folders & sub folder tests (eg duplicate looking but original sub folder paths)
	chroot
	files going to the correct folders
	multi file transfer
	etc

**Known issues:**

	o Some classes such as CmdRNFR and couple or so of others are not implemented with this code so certain uses could fail to work properly.
	
	o Android 12 is not allowing files to be created in dirs with eg "?" in the name but allows this with "#" and also fine with creating dir with "?" in the name.
	
	o At least on Android 12 with the new code, should eg FileZilla have a set remote path in site settings that is different from the app, then the new code currently
	  doesn't deal with this where the input param is blatantly incorrect.

	o When reinstalling app after having used write external storage, the app suggests its fine via the storage path, but it is not. Write external storage must be
	  used again or the code will fail to work. Hardened new code here against hard crashes but that's it.

**Random notes:**

	o Older storage code mostly kept the same and so older Android was tested for correct code execution against the new code.
	    Two exceptions to this are:
	    	o LIST code improvement is notable change with it otherwise seen as working fine.
	    	o Added code for older Android versions to use the new storage where possible to get the app working where it would otherwise fail to work at all.
	
	o Android 7.1.1 sd card didn't need the new code as it could read and write to the sd card with the old code.
	
	o FileZilla has a setting to increase its timeout so eg LIST failure due to amount of files can be further fixed by increasing that (is up to client application).
	    But, the new LIST optimizations greatly decreases the time so at least its much better overall.

	o Noting the device battery optimization to be complete. Eg Samsung needs Unrestricted to be set or connection failures will happen as the app gets killed (it will
	    happen even in the middle of use making it seem like the app code has a problem when it doesn't).

Xavron commented:
To note. I have some raw use in there for minimizing duplication. I'm sure that will totally freak out some people rofl. Its easy and quick enough to create duplication from it anyway if its not palatable. :)

Just briefly tested clean build of the code pull so its all there and working.

Possible someone has a device where some issues show up that I've not seen.

Also did not run Firebase auto testing, monkey test, or other auto tests on this which I normally do for weeks.

Xavron commented:
One more thing I forgot to add. I intentionally "separated" the scoped code from what you had to not introduce any new issues and maintain backward compatibility as it is currently. Could then work on either as needed instead of dealing with even more conflicts and regressions.

Since File isn't compatible with DocumentFile and scoped storage, I felt this was best. Didn't want outright full duplication so eliminated some of the worst spots for that.

ppareit commented:
Great! Will look into getting this pulled in.

ppareit commented:
> One more thing I forgot to add. I intentionally "separated" the scoped code from what you had to not introduce any new issues and maintain backward compatibility as it is currently. Could then work on either as needed instead of dealing with even more conflicts and regressions.

Glad you did that!

Xavron commented:
Now also briefly tested as good on Android 13 as well. It should be all good there anyway but now I got some actual use in :)

Xavron commented:
Running new different application and seeing an app crash. Will likely look into it and tack that on.

Xavron commented:
https://github.com/ppareit/swiftp/blob/6bb7a94739e24b7d0ae1d115cb27c84002d8c5b3/app/src/main/java/be/ppareit/swiftp/server/CmdMKD.java#L72

I didn't expect that. Got a FTP client application rejecting the 250 but accepts the 257. The 257 is also used by a domain name registrar. Although, I see some seem to suggest that 250 can be used or both even lol.

Kind of something like this: https://stackoverflow.com/questions/46201096/handling-babyftp-mkd-250-response-with-indy

Any thoughts on that if you're not too busy?

eakteam commented:
It seems not to be working on Android 13...
I just forked from this repo -> https://github.com/Xavron/swiftp/tree/Scoped_storage and tried to test it.
Installed on my phone Samsung Note 20 but it fails to select starting directory if writing external storage is enabled. Also it fails to list folder and files sometimes and shows empty.

Tested with FileZilla

Xavron commented:
> @Xavron It seems not to be working on Android 13... I just forked from this repo -> https://github.com/Xavron/swiftp/tree/Scoped_storage and tried to test it. Installed on my phone Samsung Note 20 but it fails to select starting directory if writing external storage is enabled. Also it fails to list folder and files sometimes and shows empty.
> 
> Tested with FileZilla

I've been primarily testing on multiple Samsung devices and including Android 12 & 13 and not sure what's happening there on your device.

While you're using "write external storage", that is fully the OS there. It simply gives access and that access is obtained and the path is then automatically entered into the apps "manage users..." section since that can't be used with scoped storage.

I've not seen the other issues you mentioned either on Samsung devices using FileZilla. If you're not using "writing external storage" than that may be a reason. The scoped storage code I added only works when "writing external storage" is in use as that is scoped storage and what is required for Android 11+. The older code has various issues I've fixed on scoped storage only.

If you have some specific error that can be traced back to the new code here then I can attempt to fix it.

Xavron commented:
> Thank you so much for your detailed explanation.
> Just in case i am using `targetSdkVersion 33`
> Regards and thanks for your great work!

You welcome.

Same here. Had it set to target 33 for a month or two now :)

Its good timing though since I just finished fixing and testing some things today. Might as well fix another and another lol.

But, still churning through the files trying to locate the cause of the last issue found. Looks like its going to take some time since its not happening often. 58 times out of 8k.

Xavron commented:
Commit 7d3424a426bec0703a764b1ad0063a5ea924e4b9

Fixed:

  o Files were misplaced or failed to transfer under a specific condition that was missed by previous testing


Tests done:

  o FileZilla transfer up of 9,828 files with 3,432 folders [SUCCESS]

  o FileZilla transfer down of 9,828 files with 3,432 folders. All there and no failures [SUCCESS]

  o All files correct and in the right place? Files transferred were source code files of a project. Downloaded project opened and build successfully [SUCCESS]

  o Two Android backups using two different client applications and a small restore test in the one [SUCCESS]

Xavron commented:
Getting a weird issue. Tested on Android 8.0 but it might affect others. Need to look into it further.

First use of write external storage results in:

```
IOException: Permission denied

IllegalArgumentException: Failed to determine if 6419-B678:backupTest/Sub/OnebackupTest/Sub/One is child of 6419-B678:backupTest/Sub/One
```

Might be a path related issue or regression since /Sub/OnebackupTest should be not that lol.

Currently fixed by force ending the app then starting it again. Its fine after that.

Xavron commented:
Possible issue (needing to be looked into):

`The upload and download of swiftp application source code using FileZilla test that I did above resulted in project having no knowledge of code changes, git log, etc.`

May be a specific issue with file bytes or very specific files in there somewhere.

Xavron commented:
Issue found:

Android 13 target of 13 FileZilla transfer down slows down to an incredible slow crawl with the screen off (even with swiftp CPU keep awake setting enabled and OS unrestricted battery enabled for the app).

Xavron commented:
> Maybe wake locks can do the job?

I thought pre-Android 13 that unrestricted battery solved that so not really sure what's going on.

However, the app already has the CPU wake lock and its still slow:

https://github.com/ppareit/swiftp/blob/6bb7a94739e24b7d0ae1d115cb27c84002d8c5b3/app/src/main/java/be/ppareit/swiftp/FsService.java#L321

Xavron commented:
[SOLVED]

> > > Maybe wake locks can do the job?

Looks like it no longer matters other than maybe to force not allow it lol.

Enabled is full wake lock and slow. Disabled is partial and fast. {SOLVED}

You were right on the pulse of it. Just the setting throws a mind bender :)

People are going to get confused on that one if the setting remains as it is.

Xavron commented:
[SOLVED] Bytes sometimes not the same on FileZilla file transfer down.

FileZilla uses ASCII which adds some bytes where FileZilla settings > Transfer > FTP: File Types Binary choice transfer equals the same amount.

So nothing to fix there.

Xavron commented:
> But it works very fine through this app -> https://github.com/wolpi/prim-ftpd on `/storage/emulated/0` so i think something is missing?

https://github.com/wolpi/prim-ftpd/blob/d0ac8a2d3a4f398df730d9f25ce9d84fad6a3386/primitiveFTPd/src/org/primftpd/PrimitiveFtpdActivity.java#L561

Normally ones not on Play Store such as this one there use all files access. That's the only way to get that exact location used now.

FYI, its no longer scoped storage when that is used. It has access to everything. That's no what scoped storage is and is more of a security risk.

Xavron commented:
> Also when adding new FTP user we are unable to change starting directory for the new user. Different users may use different starting location ?

This one is that you pick the target location using write external storage. Then in each client app you set the path you want.

Eg:

```
write external storage set to "/storage/emulated/0/ftpFiles/"

Client A: uses path "/a"

Client B: uses path "/b"
```

or it could be:

```
write external storage set to "/storage/emulated/0/ftpFiles/"

Client A: uses path ""

Client B: uses path ""

and they both use the same location in that case.
```

Xavron commented:
> This video shows that even when enabling `Write External Storage` i'm unable to list folders/files inside DCIM

Unfortunately, I can't reproduce it.

Scoped storage set to "/storage/emulated/0/DCIM/" gives that path back with permission. Its the exact path on all Samsung devices and works on my test Android 13 device immediately.

I can't even think of any reason why it wouldn't produce files there :'(

Xavron commented:
> > > This video shows that even when enabling `Write External Storage` i'm unable to list folders/files inside DCIM
> > 
> > 
> > Unfortunately, I can't reproduce it.
> > Scoped storage set to "/storage/emulated/0/DCIM/" gives that path back with permission. Its the exact path on all Samsung devices and works on my test Android 13 device immediately.
> > I can't even think of any reason why it wouldn't produce files there :'(
> > _It could be possible that your FileZilla is dirty. Changing write external storage can cause FileZilla to mess up the paths. I have that noted way above. When you change write external storage you need to exit FileZilla. If it was a saved site then you'd also have to make sure the paths in there were correct too but it looks like you're using quick connect so only the former could be an issue. Although testing this now again and am not seeing an issue that way._
> 
> I'm able to reproduce the issue. So:
> 
>     1. Enable Write External Storage to any folder and start FTP Server.
> 
>     2. Connect via FileZilla (Directory listing successful/file listing successful).
> 
>     3. Grab anything from the PC (Video, Picture etc. and upload it to the opened folder.
> 
>     4. File upload successfully and also exist on the phone storage via its file manager `BUT`
> 
> 
> After that FileZilla is unable to list folder and files again to the same folder where we upload the file from PC. Is weird, but that's the steps to reproduce it :)

Thank you. But sadly, its not enough. I see the paths duplicating in your video but have no idea why.

It could be new a variation needing to be dealt with in inputPathToChrootedFile() but there's not enough to go on unless I just randomly upload test code and I can't do that here lol.

Xavron commented:
Alright. You have an MSLD error in the video. That does show up just before a listing of a directory. So does files to list, get working directory variable, violation of chroot, and a while lot more that could cause that failure there.

There's just not enough to go on. I have to be able to run it through the debugger.

Or you need to provide what's happening in logcat.

Xavron commented:
> 

I was expecting that to help but its only showing MLSD is fine there :\

Until I can reproduce it, I can't do anything to fix it. Sorry :(

Can't do certain other things here because of the branch being a pull request.

Xavron commented:
> Any idea how can i debug it to find the root cause of why is this happening?

Thanks :) But, I just reproduced it as well. Had to push a file to DCIM folder to do it. Its not done it with any other folder so I'm wondering if its the app or Android thing.

Have to debug it... will get back once I do that and determine what it actually is :)

Xavron commented:
Commit f3bb8972625f9a1047669b5aab6ae4399353f49c

Fixed:

`Some dirs getting empty file listings especially after transferring a file`

Tests:

```
Internal DCIM folder and pushing files to it.

Tested for conflicts briefly with sd card and two client applications. Not doing extensive testing on this one.
```

Xavron commented:
> Another Bug(maybe).
> 
> Everything seems to works fine except that after uploading something to a directory it also creates a folder named `storage` after uploading is successful. Inside that directory is another one `emulated` and another one inside `0`.

Yeah, its all path stuff. Technically, its all fixed now (without conflict or further extensive testing anyway) but won't be committing it for some time because I no longer like the code and need to test a lot. That will take a while to properly do.

Xavron commented:
commit 47efb9c3252f678ebf262abbd8d66a3ccd1d07b5

Simplified path code with return to some older code that's now working fine with it after things have been fixed in past commits and a couple more in this one. Didn't see that one coming.

```
Known issues not fixed:

  1. Ftp client already connected and then switching write external storage location causes chroot to not get updated.

      To reproduce:

        o Create and enter a dir, use write external storage and change locations eg sd card to internal, refresh, create dir and old chroot is still in use.

      Notes:

        o What happens on multiple clients connected? Disconnect them all?

        o Hot swap chroot location on all or temporarily keep the last UriPermission around until all clients finish?

  2. Seen the app hard crashing during debugging when FileZilla connects.

      o Not the app. Debugger issue. One workaround is by connecting without debugger attached then attach after.

```
```
Android 13:

  o Backups continual sd card two backups using two different clients [x]
  o Backups full restore over 25GB total sd card two backups using two different clients []

  o Multiple clients connected and transferring at the same time [x]

  o FileZilla:

    transfer tests up & down:

      swiftp source code transfer up and down plus clean and build:
        sd card /dir/1/ [x]
        local /DCIM/ [x] 
        local /Test/ [x]

      local /DCIM/ one file at chroot [x]
      local /Test/ one file at chroot [x]

    sd card:
      transfer up recursive [x]
      transfer down recursive [x]
      transfer up one  file [x]
      transfer down one  file [x]
      delete recursive [x]
      delete one [x]
      movement one dir up [x]
      movement one dir down [x]
      movement skip multi dirs [x]
      new dir [x]
      dir refresh [x]
      dir rename [x]
      9.8k files + 3.4k dirs u [x]
      9.8k files + 3.4k dirs d [x]

    local:
      transfer up recursive [x]
      transfer down recursive [x]
      transfer up one file [x]
      transfer down one file [x]
      delete recursive [x]
      delete one [x]
      movement one dir up [x]
      movement one dir down [x]
      movement skip multi dirs [x]
      new dir [x]
      dir refresh [x]
      9.8k files + 3.4k dirs u [x]
      9.8k files + 3.4k dirs d [x]

  ________________
  ALL SUCCESSFUL WITH THIS COMMIT - Files transferred over 47k times successfully in total on local and sd card.
```

ppareit commented:
Lot of internal changes! Scoped storage will be the way forward.

ppareit commented:
Also made some little changes to setup to make Android 13 work out of the box. See `PreferenceFragment.onCreate`.

Xavron commented:
I'll be continuing to monitor for any unforeseen issues and fix them as I can.
